modules/ITK/cli/utils/statismo-build-gp-model-kernels.h

```diff
@@ -95,12 +95,181 @@ class GaussianKernel : public statismo::ScalarValuedKernel<TPoint> {
         return os.str();
     }
 
-  private:
+private:
     double m_sigma;
     double m_sigma2;
 };
 
 
+
+
+
+template <class TPoint>
+class BSplineKernel : public statismo::ScalarValuedKernel<TPoint> {
+public:
+    typedef typename  TPoint::CoordRepType CoordRepType;
+    typedef vnl_vector<CoordRepType> VectorType;
+
+
+    BSplineKernel(double support) : m_support(support) {}
+
+
+    // the 3rd order b-spline basis function
+    double bspline3(const double& x) const {
+        const double absX = std::fabs(x);
+        const double absXSquared = absX * absX;
+        const double absXCube = absXSquared * absX;
+        const double twoMinAbsX = 2.0 - absX;
+        const double twoMinAbsXCube = twoMinAbsX * twoMinAbsX * twoMinAbsX;
+        const double twoByThree = 2.0 / 3.0;
+
+        double splineValue = 0;
+        if (absX >= 0 && absX < 1) {
+            splineValue = twoByThree - absXSquared + 0.5 * absXCube;
+        }
+        else if (absX >= 1 && absX < 2) {
+            splineValue = twoMinAbsXCube / 6.0;
+        }
+        else {
+            splineValue = 0;
+        }
+        return splineValue;
+    }
+
+    double tensorProductSpline(const VectorType& x) const {
+        double prod = 1;
+        for (unsigned d = 0; d < x.size(); ++d) {
+            prod *= bspline3(x[d]);
+        }
+        return prod;
+    }
+
+
+    inline double operator()(const TPoint& x, const TPoint& y) const {
+
+        assert(x.Size() == y.Size());
+        const unsigned dim = x.Size();
+
+        if (dim < 1 || dim > 3) {
+            std::ostringstream os;
+            os << "Currently only dimensions 1 - 3 are supported.  ( Received" << dim << ")";
+            throw statismo::StatisticalModelException(os.str().c_str());
+        }
+
+        const double order = 3;
+        const double supportBasisFunction = 4.0;
+        const double scale = -1.0 * std::log(m_support / supportBasisFunction) / std::log(2);
+
+        VectorType xScaled = x.GetVnlVector() * std::pow(2.0, scale);
+        VectorType yScaled = y.GetVnlVector() * std::pow(2.0, scale);
+
+        VectorType kLower(dim);
+        VectorType kUpper(dim);
+        for (unsigned d = 0; d < dim; ++d) {
+            kLower[d] = static_cast<int>(std::ceil(std::max(xScaled[d], yScaled[d]) - 0.5 * supportBasisFunction));
+            kUpper[d] = static_cast<int>(std::floor(std::min(xScaled[d], yScaled[d]) + 0.5 * supportBasisFunction));
+        }
+
+
+        // We need to generate the cartesian product k_1 x ... x k_d, where k_i goes through all the integers
+        // within the given bounds. A non-recursive solution requires d loops. Here we just write down the cases
+        // for 1 2 and 3D
+        double sum = 0.0;
+        double kx = kLower[0];
+        while (kx <= kUpper[0]) {
+            if (dim == 1) {
+                VectorType k(1);
+                k[0] = kx;
+                sum += (tensorProductSpline(xScaled - k) * tensorProductSpline(yScaled - k));
```

Comment:
yes, for the 1d case this would be possible. But I think it makes the structure of the loop even less clear. (I already don't like the solution with the dependence on the dimension, but I found it the most pragmatic choice). I therefore prefer to call tensorProductSpline even in the 1D case (performance in the 1D case is not critical).